lib/collection/src/common/snapshots_manager.rs

```diff
@@ -9,34 +11,83 @@ use crate::common::sha_256::hash_file;
 use crate::operations::snapshot_ops::{
     get_checksum_path, get_snapshot_description, SnapshotDescription,
 };
-use crate::operations::types::CollectionResult;
+use crate::operations::snapshot_storage_ops;
+use crate::operations::types::{CollectionError, CollectionResult};
+use crate::shards::shard::ShardId;
+use crate::shards::shard_holder::LockedShardHolder;
 
-#[derive(Clone, Deserialize, Debug)]
+#[derive(Clone, Deserialize, Debug, Default)]
+pub struct SnapShotsConfig {
+    pub snapshots_storage: SnapshotsStorageConfig,
+    pub s3_config: Option<S3Config>,
+}
+
+#[derive(Clone, Debug, Default, Deserialize)]
+pub enum SnapshotsStorageConfig {
+    #[default]
+    Local,
+    S3,
+}
+
+#[derive(Clone, Deserialize, Debug, Default)]
 pub struct S3Config {
     pub bucket: String,
     pub region: Option<String>,
     pub access_key: Option<String>,
     pub secret_key: Option<String>,
+    pub endpoint_url: Option<String>,
 }
 
 #[allow(dead_code)]
-pub struct SnapshotStorageS3 {
-    s3_config: S3Config,
+pub struct SnapshotStorageCloud {
+    client: Box<dyn object_store::ObjectStore>,
 }
 
 pub struct SnapshotStorageLocalFS;
 
 pub enum SnapshotStorageManager {
     LocalFS(SnapshotStorageLocalFS),
-    S3(SnapshotStorageS3),
+    // Assuming that we can have common operations for all cloud storages
+    S3(SnapshotStorageCloud),
+    // <TODO> : Implement other cloud storage
+    // GCS(SnapshotStorageCloud),
+    // AZURE(SnapshotStorageCloud),
```

Comment:
Assuming that we can use `SnapshotStorageCloud` for all.
If it's not possible, we need to switch back to implementing`SnapshotStorageS3` or `SnapshotStorageGCS` for each.